framework-developer/plans/gmail-oauth-authentication.md

```diff
@@ -0,0 +1,1731 @@
+            ResponseEntity<AccountQueryModelState> response = restTemplate.exchange(
+                "/v1/accounts/{userId}",
+                HttpMethod.GET,
+                entity,
+                AccountQueryModelState.class,
+                userId
+            );
+            return response.getBody();
+        } catch (HttpClientErrorException.Unauthorized ex) {
+            // Refresh token and retry
+            refreshAccessToken();
+            return getAccount(userId);
+        }
+    }
+    
+    private void refreshAccessToken() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        HttpEntity<RefreshTokenRequest> entity = new HttpEntity<>(
+            new RefreshTokenRequest(refreshToken),
+            headers
+        );
+        
+        ResponseEntity<TokenResponse> response = restTemplate.postForEntity(
+            "/auth/refresh",
+            entity,
+            TokenResponse.class
+        );
+        
+        this.accessToken = response.getBody().accessToken();
+    }
+}
+```
+
+**Security Considerations for JWT:**
+
+1. **GCP Service Account Key Management:**
+   - Store service account JSON key in Kubernetes secret, never in code or repository
+   - Restrict service account permissions (only needs JWT signing capability)
+   - Monitor service account usage via GCP Cloud Audit Logs
+   - Rotate service account keys periodically (Google recommends every 90 days)
+   - Use Workload Identity in GKE for key-less authentication (advanced)
+
+2. **JWKS Endpoint Security:**
+   - JWKS endpoint is public (only contains public keys, safe to expose)
+   - Queries service caches JWKS keys to minimize network calls
+   - Spring Security automatically refreshes keys when new `kid` (Key ID) is encountered
+   - Google handles key rotation automatically, supports multiple active keys
+
+3. **Token Storage:**
+   - Client should store tokens in memory or secure storage
+   - Avoid localStorage for browser apps (XSS risk)
+   - For server-to-server communication, store in environment or secret manager
+   - Tokens are self-contained, no server-side session required
+
+4. **Token Validation:**
+   - Validate RS256 signature using JWKS public keys on every request
+   - Check expiration time (`exp` claim)
+   - Verify issuer (`iss` claim) matches expected value
+   - Verify audience (`aud` claim) matches your API identifier
+   - Implement token revocation list for logout (optional, requires Redis or database)
+
+5. **HTTPS Only:**
+   - All token transmission must use HTTPS in production
+   - GCP Service Account keys must never be transmitted over HTTP
+   - OAuth redirect URIs must use HTTPS (except localhost for development)
+
+#### 6. OAuth2 User Service
+
+Custom `OAuth2UserService` implementation:
+1. Receives OAuth2 user info from provider
+2. Extracts user details (email, name, provider ID)
+3. Checks if account exists (by email or provider ID)
```

Comment:
This implies it needs to have a Read model available. Use the example of the AccountQueryModel in the queries module.